provider/openstack/networking.go

```diff
@@ -202,12 +167,68 @@ func (n *NeutronNetworking) AllocatePublicIP(instId instance.Id) (*string, error
 	return nil, lastErr
 }
 
+// getExternalNetworkIDs returns a slice of external network IDs.
+// If specified, the configured external network is returned.
+// Otherwise search for an external network in the same
+// availability zone as the private network.
+func (n *NeutronNetworking) getExternalNetworkIDs() ([]string, error) {
+	extNetworkIds := make([]string, 0)
+	neutronClient := n.env.neutron()
+	externalNetwork := n.env.ecfg().externalNetwork()
+	if externalNetwork != "" {
+		// the config specified an external network, try it first.
+		netId, err := resolveNeutronNetwork(neutronClient, externalNetwork, true)
+		if err != nil {
+			logger.Debugf("external network %s not found, search for one", externalNetwork)
+		} else {
+			logger.Debugf("using external network %q", externalNetwork)
+			extNetworkIds = []string{netId}
+		}
+	}
+
+	if len(extNetworkIds) == 0 {
+		// Create slice of network.Ids for external networks in the same AZ as
+		// the instance's network, to find an existing floating ip in, or allocate
+		// a new floating ip from.
+		configNetwork := n.env.ecfg().network()
+		netId, err := resolveNeutronNetwork(neutronClient, configNetwork, false)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		netDetails, err := neutronClient.GetNetworkV2(netId)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+
+		availabilityZones := netDetails.AvailabilityZones
+		if len(availabilityZones) == 0 {
+			// No availability zones is valid, check for empty string
+			// to ensure we still find the external network with no
+			// AZ specified.  lp: 1891227
+			availabilityZones = []string{""}
+		}
+		for _, az := range availabilityZones {
+			extNetIds, _ := getExternalNeutronNetworksByAZ(n.env, az)
+			if len(extNetIds) > 0 {
+				extNetworkIds = append(extNetworkIds, extNetIds...)
+			}
+		}
+
+		if len(extNetworkIds) == 0 {
+			return nil, errors.NotFoundf(
+				"could not find an external network in availability zone %s", netDetails.AvailabilityZones)
```

Comment:
this error message could use some tweaking.
1. it is a iist of zones and messages says "zone"
2. for the az="" case (the bug fix), the message doesn't seem quite right